provisioner/terraform/tfparse/tfparse.go

```diff
@@ -0,0 +1,490 @@
+package tfparse
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"os"
+	"slices"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/coder/coder/v2/archive"
+	"github.com/coder/coder/v2/provisionersdk"
+	"github.com/coder/coder/v2/provisionersdk/proto"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclparse"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/terraform-config-inspect/tfconfig"
+	"github.com/zclconf/go-cty/cty"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+)
+
+// NOTE: This is duplicated from coderd but we can't import it here without
+// introducing a circular dependency
+const maxFileSizeBytes = 10 * (10 << 20) // 10 MB
+
+// WorkspaceTags extracts tags from coder_workspace_tags data sources defined in module.
+// Note that this only returns the lexical values of the data source, and does not
+// evaluate variables and such. To do this, see evalProvisionerTags below.
+func WorkspaceTags(ctx context.Context, logger slog.Logger, module *tfconfig.Module) (map[string]string, error) {
+	workspaceTags := map[string]string{}
+
+	for _, dataResource := range module.DataResources {
+		if dataResource.Type != "coder_workspace_tags" {
+			logger.Debug(ctx, "skip resource as it is not a coder_workspace_tags", "resource_name", dataResource.Name, "resource_type", dataResource.Type)
+			continue
+		}
+
+		var file *hcl.File
+		var diags hcl.Diagnostics
+		parser := hclparse.NewParser()
+
+		if !strings.HasSuffix(dataResource.Pos.Filename, ".tf") {
+			logger.Debug(ctx, "only .tf files can be parsed", "filename", dataResource.Pos.Filename)
+			continue
+		}
+		// We know in which HCL file is the data resource defined.
+		file, diags = parser.ParseHCLFile(dataResource.Pos.Filename)
+		if diags.HasErrors() {
+			return nil, xerrors.Errorf("can't parse the resource file: %s", diags.Error())
+		}
+
+		// Parse root to find "coder_workspace_tags".
+		content, _, diags := file.Body.PartialContent(rootTemplateSchema)
+		if diags.HasErrors() {
+			return nil, xerrors.Errorf("can't parse the resource file: %s", diags.Error())
+		}
```

Comment:
I think I just lack the context for how this works. The input to this function is `*tfconfig.Module`, which I thought was a singular resource block.

Does this mean we parse the same file for each resource block it has inside of it? So if a file has 3 blocks, we call `parser.ParseHCLFile("somefile.tf")` 3 times?

Reply:
Yes, because `tfconfig.Module` contains already processed forms, and here we need to raw form with original variables hence ... one more parsing.

Reply:
@mtojek can we cache the previously parsed? Or is this quick enough it does not really matter?

Reply:
I think so, it is matter of implementing following the clean code rules, but should be doable!

Reply:
Turns out that `hclparse.Parser()` will cache by default. I'm wary of keeping a single instance around forever but we can definitely re-use a single parser for a single instance of parsing a given Terraform configuration.